aten/src/ATen/native/TensorIndexing.cpp

```diff
@@ -154,5 +152,189 @@ std::ostream& operator<<(std::ostream& stream, const std::vector<TensorIndex>& t
   return stream;
 }
 
+// This mirrors `count_specified_dimensions` in torch/csrc/autograd/python_variable_indexing.cpp
+int64_t count_specified_dimensions(ArrayRef<TensorIndex> indices) {
+  // Count the number of indexed dimensions (everything but ellipsis and None)
+  int64_t count = 0;
+  size_t size = indices.size();
+  for (size_t i = 0; i < size; i++) {
+    auto& obj = indices[i];
+    if (obj.is_tensor()) {
+      auto& tensor = obj.tensor();
+      if (tensor.scalar_type() == kByte || tensor.scalar_type() == kBool) {
+        count += tensor.dim();
+      } else {
+        count++;
+      }
+    } else if (!obj.is_none() && !obj.is_ellipsis() && !obj.is_boolean()) {
+      count++;
+    }
+  }
+  return count;
+}
+
+// This mirrors `applySlicing` in torch/csrc/autograd/python_variable_indexing.cpp
+Tensor applySlicing(const Tensor& self, ArrayRef<TensorIndex> indices, std::vector<Tensor>& outIndices) {
```

Comment:
This code doesn't correspond all that closely to the version in `python_variable_indexing`. I can believe that they are the same, but it's non obvious how one would port changes from one end to the other. It would be easier to review this code if the structure more exactly lined up.